Update API conventions. Add kubectl conventions.

bgrant0607 opened the pull request:
Ref #12322. Fixes #6797.

cc @smarterclayton @erictune @thockin

bgrant0607 committed:
Update API conventions. Add kubectl conventions.
Ref #12322. Fixes #6797.

ncdc commented:
/cc @jwforres

cjcullen commented:
@k8s-bot test this

smarterclayton commented on docs/devel/api-conventions.md:
```diff
@@ -172,18 +183,38 @@ Objects that contain both spec and status should not contain additional top-leve
 
-* **phase**: The phase is a simple, high-level summary of the phase of the lifecycle of an object. The phase should progress monotonically. Typical phase values are `Pending` (not yet fully physically realized), `Running` or `Active` (fully realized and active, but not necessarily operating correctly), and `Terminated` (no longer active), but may vary slightly for different types of objects. New  ...
-* **conditions**: Conditions represent orthogonal observations of an object's current state. Objects may report multiple conditions, and new types of conditions may be added in the future. Condition status values may be `True`, `False`, or `Unknown`. Unlike the phase, conditions are not expected to be monotonic -- their values may change back and forth. A typical condition type is `Ready`, which  ...
+**Conditions** represent the latest available observations of an object's current state. Objects may report multiple conditions, and new types of conditions may be added in the future. Therefore, conditions are represented using a list/slice, where all have similar structure.
```
Describe why we don't use phase?  It's an important distinction

bgrant0607 replied:
Added below.

smarterclayton commented on docs/devel/api-conventions.md:
```diff
@@ -423,6 +442,9 @@ APIs may return alternative representations of any resource in response to an Ac
 
+## Time Field Names
+
+The name of a field that specifies the time at which `something` occurs should be called `somethingTime`. Do not use `stamp` (e.g., `creationTimestamp`).
```
So our existing meta fields should be changed in v2?

bgrant0607 replied:
We use at least 3 different conventions in the API currently. I picked the most aesthetically pleasing one for new fields. Whether it's worth changing existing fields is another issue.

smarterclayton commented on docs/devel/kubectl-conventions.md:
```diff
@@ -0,0 +1,115 @@
+  * `-Lkey` should append a column containing the value of label with key `key`, with `<none>` if not present
+  * json, yaml, Go template, and jsonpath template formats should be supported and encouraged for subsequent processing
+    * Users should use --api-version or --output-version to ensure the output uses the version they expect
+* `describe` commands may output on multiple lines and may include information from related resources, such as events
```
Add: Describe should add additional information from related resources that a normal user may need to know - if a user would always run "describe resource1" and the immediately want to run a "get type2" or "describe resource2", consider including that info.  Examples, persistent volume claims for pods that reference claims, events for most resources, nodes and the pods scheduled on them.

derekwaynecarr replied:
When fetching related resources, a targeted field selector should be used in favor of client side filtering of related resources.  

Have been fixing a few of our describe calls that fetch everything instead of a more targeted call.

bgrant0607 replied:
Used text verbatim. Thanks.

smarterclayton commented:
Few minor, I'll try to read through and suggest other things by looking through my issues list in a follow on.

derekwaynecarr commented on docs/devel/api-conventions.md:
```diff
@@ -117,9 +121,15 @@ Kinds are grouped into three categories:
+   * `/binding`: Used to bind a resource representing a user request (e.g., Pod, PersistentVolumeClaim) to a cluster infrastructure resource (e.g., Node, PersistentVolume).
+   * `/status`: Used to write just the status portion of a resource. For example, the `/pods` endpoint only allows updates to `metadata` and `spec`, since those reflect end-user intent. An automated process should be able to modify status for users to see by sending an updated Pod kind to the server to the "/pods/&lt;name&gt;/status" endpoint - the alternate endpoint allows different rules to be  ...
+   * `/scale`: Used to read and write the count of a resource in a manner that is independent of the specific resource schema.
+
```
We also have a /finalize resource but that may prove to be less common.

bgrant0607 replied:
I'll add that in a later pass. I hope to find time to hash out initializers in a couple weeks.

bgrant0607 commented:
Addressed feedback. I squashed because I to rebase and force push in order to undo an accidental change to workaround the need-too-much-memory-to-build problem.

smarterclayton commented:
No further comments, LGTM

On Thu, Aug 13, 2015 at 1:15 AM, Brian Grant <notifications@github.com>
wrote:

> Reopened #12515 <https://github.com/kubernetes/kubernetes/pull/12515>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12515#event-381075151>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

bgrant0607 commented:
Rebased to see if that fixes tests.
